### DISCUSSION TITLE
I think I figured out how recording and retention works in Frigate

### DISCUSSION THREAD

**Discussion by ndbroadbent** [blakeblackshear/frigate#4231] [General]
I set up Frigate yesterday and left a camera on the floor overnight to try it out. I was a bit confused about how the recording works, but I think I've figured it out now. I thought I would post here in case it helps someone else, or in case I still have some details wrong.

I use the following `record` config:

```
record:
  enabled: True
  retain:
    days: 7
    mode: motion
  events:
    retain:
      default: 14
      mode: active_objects
```

### Main Frigate page showing cameras

<img width="1059" alt="Screen Shot 2022-11-02 at 2 25 18 PM" src="https://user-images.githubusercontent.com/139536/199372920-32b69a79-577e-4776-9a23-bb375fc6bbdf.png">

### Events page

This is related to this part of the config:

```
  events:
    retain:
      default: 14
      mode: active_objects
```

<img width="1289" alt="Screen Shot 2022-11-02 at 2 25 40 PM" src="https://user-images.githubusercontent.com/139536/199372953-67ad054a-a6a6-4d2c-a042-9ef36c8fd2a4.png">

It only saves video clips when an object is detected, and the object is moving.

### Specific event

<img width="1659" alt="Screen Shot 2022-11-02 at 2 27 04 PM" src="https://user-images.githubusercontent.com/139536/199373112-0c257faf-471c-4a99-af87-d5f92ba56bd3.png">

This shows the specific recorded clip

### Recordings

This is related to this part of the config:

```
  retain:
    days: 7
    mode: motion
```

This is similar to the events page, but instead of showing a trimmed clip, it actually shows the full recorded video for that hour, and links to specific timestamps in that recording video. You can also click on events to go to that timestamp.

<img width="1664" alt="Screen Shot 2022-11-02 at 2 28 44 PM" src="https://user-images.githubusercontent.com/139536/199373284-b0523e6f-c3c5-4d9e-ac04-90bb8c73bf2e.png">

You can also click on a time to go to the beginning of that hour, so you can see any motion as well (not just events):

<img width="397" alt="Screen Shot 2022-11-02 at 2 29 40 PM" src="https://user-images.githubusercontent.com/139536/199373480-0baf0b82-5ad7-4ee4-be02-2e9498761b5d.png">

The recording will only show any 10 second segments of video where motion was detected. Each 10 second segment of motion is merged together into a single video file for that hour: `Recordings/MM YYYY/MM DD/HH:MM:SS/<camera name>`
So your recording for that hour might be only a few seconds or minutes long, if not much motion happened.

### Home Assistant Media folder

<img width="989" alt="Screen Shot 2022-11-02 at 2 27 32 PM" src="https://user-images.githubusercontent.com/139536/199373150-9e708631-54c9-4f69-91b1-230b256ab01a.png">

* Clips = Video clips recorded for Frigate events (`active_objects`)
* Recordings = One video file for every hour, where any motion was detected. Organized in folders: `Recordings/MM YYYY/MM DD/HH:MM:SS/<camera name>`

<img width="670" alt="Screen Shot 2022-11-02 at 2 32 48 PM" src="https://user-images.githubusercontent.com/139536/199373816-a7f6af85-c141-49a2-993f-03da1193bf95.png">

<img width="1663" alt="Screen Shot 2022-11-02 at 2 33 07 PM" src="https://user-images.githubusercontent.com/139536/199373852-8ba51ed3-1ed3-45dd-b9af-24bb0052133a.png">

I was looking at a recording from 3am last night:

<img width="1298" alt="Screen Shot 2022-11-02 at 2 35 01 PM" src="https://user-images.githubusercontent.com/139536/199374046-c1a118b9-d0a6-49c0-ab2a-530267c71f8d.png">

It looks like Frigate decided to keep 11:48 of video from the hour. I couldn't figure out why, since I didn't see any motion in the video. Then I realized it was probably detecting motion from the embedded timestamp overlay that I had configured in the IP camera settings.
 
<img width="189" alt="Screen Shot 2022-11-02 at 2 36 54 PM" src="https://user-images.githubusercontent.com/139536/199374217-6c918991-e00c-4315-b6e9-f16ce893923e.png">

I guess it was usually ignoring the seconds, but it detected enough of a difference when it ticked over to the next minute: `02:06:59` => `02:07:00`

It looks like Frigate can't (or shouldn't) add these timestamps itself: https://github.com/blakeblackshear/frigate/issues/2151
(Or maybe you could do it with ffmpeg if you really need it.) But I want to keep the timestamp, so it looks like I need a motion mask to cover up the timestamp area: https://docs.frigate.video/configuration/masks

<img width="1661" alt="Screen Shot 2022-11-02 at 2 52 56 PM" src="https://user-images.githubusercontent.com/139536/199376009-4cdfdc25-a134-401f-8e68-6745ab8f6b97.png">

```
cameras:
  living_room_camera:
    ffmpeg:
      inputs: ...
      output_args: ...
    motion:
      mask:
        - 1000,24,997,46,678,47,677,24        
    detect:
      width: 1024
      height: 576
      fps: 5
```

-------

I think that's everything I've learned so far. I was worried that Frigate would only record videos if it detected an object. I was concerned that there would be a false negative where someone breaks in and steals something but we didn't get any video because the person detection failed for some reason. So it's great to know that it does actually record video for any motion, and it also saves additional clips for any object detection events.

**Comment by NickM-27** [comment]
From what I can see mostly correct except for one detail. When viewing an event the event will gather ALL segments relevant to that event. So if since you have 24/7 recordings enabled with motion based retention. Any segments with motion during that event will also be shown as part of that event (until the 7 days you set are up at which point the event will only show the active_object segments)

**Reply by ndbroadbent** [reply]
Thanks @NickM-27! Do you mean on the Events page, or when you click an event on the Recordings page?

If I click an event on the Events page, is that not just pulling up the clip from the Clips folder?
 
<img width="1722" alt="Screen Shot 2022-11-02 at 3 17 14 PM" src="https://user-images.githubusercontent.com/139536/199379790-78b2d120-27a0-41f5-9524-4c7ff01ba0d9.png">

This is just showing me the 20 second clip where motion was detected for an object.

Oh or is this for when you have multiple cameras? (I just have one for now.)

**Reply by NickM-27** [reply]
There are no clips in the `clips` folder, only snapshots. Both events and recordings playback are generated from the ~10 second segments. 

You can read https://github.com/blakeblackshear/frigate/issues/4221#issuecomment-1298889852 where I explain this in more detail

**Reply by ndbroadbent** [reply]
Ahhhh I see! Thanks for that comment, that's very helpful. I also took a closer look at the `media` directory, and it makes sense now. I can see the individual files for clips, and how they're being stitched together in the Frigate UI and Home Assistant Media browser.

```
/usr/share/hassio/media/frigate# ls -l
total 5284
drwxr-xr-x 2 root root    4096 Nov  1 17:37 clips
-rw-r--r-- 1 root root 1224704 Nov  2 15:25 frigate.db
-rw-r--r-- 1 root root   32768 Nov  2 15:47 frigate.db-shm
-rw-r--r-- 1 root root 4136512 Nov  2 15:47 frigate.db-wal
drwxr-xr-x 3 root root    4096 Nov  1 18:57 recordings

/usr/share/hassio/media/frigate# ls recordings/2022-11/02/14/living_room_camera/
00.05.mp4  02.05.mp4  04.04.mp4  05.54.mp4  07.04.mp4  08.54.mp4  11.54.mp4  13.04.mp4  14.54.mp4  16.04.mp4  19.01.mp4  23.02.mp4  25.02.mp4  28.02.mp4  30.02.mp4  34.01.mp4  37.01.mp4  40.01.mp4  44.01.mp4  47.01.mp4  50.01.mp4  54.01.mp4  57.01.mp4  59.11.mp4
00.15.mp4  02.54.mp4  04.54.mp4  06.04.mp4  07.54.mp4  09.54.mp4  12.04.mp4  13.54.mp4  15.04.mp4  16.54.mp4  20.01.mp4  23.12.mp4  26.02.mp4  29.02.mp4  32.02.mp4  35.01.mp4  38.01.mp4  42.01.mp4  45.01.mp4  48.01.mp4  52.01.mp4  55.01.mp4  58.01.mp4  59.21.mp4
01.55.mp4  03.54.mp4  05.04.mp4  06.54.mp4  08.04.mp4  10.04.mp4  12.54.mp4  14.04.mp4  15.54.mp4  17.04.mp4  22.02.mp4  24.02.mp4  27.02.mp4  29.52.mp4  33.01.mp4  36.01.mp4  39.01.mp4  43.01.mp4  46.01.mp4  49.01.mp4  53.01.mp4  56.01.mp4  59.01.mp4
```

Thanks!

**Comment by ndbroadbent** [comment]
Whoops, this `motion` mask config didn't work. I think I might have put it in the wrong place?

```
cameras:
  living_room_camera:
    ffmpeg:
      inputs: ...
      output_args: ...
    motion:
      mask:
        - 1000,24,997,46,678,47,677,24  
```

I just looked at the parsed config on the debug page

```
"cameras": {
    "living_room_camera": {
...
      "motion": {
        "contour_area": 30,
        "delta_alpha": 0.2,
        "frame_alpha": 0.2,
        "frame_height": 50,
        "improve_contrast": false,
        "mask": "",
        "mqtt_off_delay": 30,
        "threshold": 25
      },
```

But then I moved it to the root of the config file and restarted, and it worked. Then I moved it back under the camera and it still worked.

I could have sworn I restarted Frigate after adding the mask. I even restarted it a few times while trying out different options, it was still stuck at `"mask": "",`. But I seem to have triggered some persistent state change somewhere (maybe in a seperate database file?) I can't reproduce the blank mask string anymore.

Oh well, it's working now!

**Comment by pipanosa** [comment]
Great!
Many thanks for your explanation, the way we can find footage with motion but no event is really not obvious. I was thinking that it was a bug...
